File: modules/core/src/main/scala/com/github/chengpohi/edql/parser/EDQLParserFactory.scala
package com.github.chengpohi.edql.parser

import com.github.chengpohi.edql.EDQLLanguage
import com.intellij.lang._
import com.intellij.psi.PsiFile
import com.intellij.psi.impl.PsiFileFactoryImpl
import com.intellij.testFramework.LightVirtualFile

import java.nio.charset.StandardCharsets

class EDQLParserFactory(val myFileExt: String,
                        val myDefinition: ParserDefinition,
                        val psiFileFactory: PsiFileFactoryImpl) {

  def createFile(name: String, text: String): PsiFile = {
    val virtualFile = new LightVirtualFile(name, EDQLLanguage.INSTANCE, text)
    virtualFile.setCharset(StandardCharsets.UTF_8)
    createFile(virtualFile)
  }

  def createFile(virtualFile: LightVirtualFile): PsiFile = {
    psiFileFactory.trySetupPsiForFile(virtualFile, EDQLLanguage.INSTANCE, true, false)
  }
}

object EDQLParserFactory {
  def apply(ext: String,
            parserDefinition: ParserDefinition,
            psiFileFactory: PsiFileFactoryImpl): EDQLParserFactory = {
    new EDQLParserFactory(ext, parserDefinition, psiFileFactory)
  }
}

File: src/main/java/com/github/chengpohi/edql/parser/EDQLParserFactory.kt
package com.github.chengpohi.edql.parser

import com.github.chengpohi.edql.EDQLLanguage
import com.intellij.lang.ParserDefinition
import com.intellij.psi.PsiFile
import com.intellij.psi.impl.PsiFileFactoryImpl
import com.intellij.testFramework.LightVirtualFile
import java.nio.charset.StandardCharsets

class EDQLParserFactory(
    val myFileExt: String,
    val myDefinition: ParserDefinition,
    val psiFileFactory: PsiFileFactoryImpl
) {

    fun createFile(name: String, text: String): PsiFile {
        val virtualFile = LightVirtualFile(name, EDQLLanguage.INSTANCE, text)
        virtualFile.charset = StandardCharsets.UTF_8
        return psiFileFactory.trySetupPsiForFile(virtualFile, EDQLLanguage.INSTANCE, false, false)!!
    }

    companion object {
        fun apply(
            ext: String,
            parserDefinition: ParserDefinition,
            psiFileFactory: PsiFileFactoryImpl
        ): EDQLParserFactory {
            return EDQLParserFactory(ext, parserDefinition, psiFileFactory)
        }
    }
} 